libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemKeyConfig.java

```diff
@@ -8,84 +8,148 @@
+import org.elasticsearch.core.Tuple;
+
 import java.io.IOException;
-import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
+import java.security.AccessControlException;
 import java.security.GeneralSecurityException;
 import java.security.KeyStore;
 import java.security.PrivateKey;
 import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.X509ExtendedKeyManager;
+
+
 /**
  * A {@link SslKeyConfig} that reads from PEM formatted paths.
  */
 public final class PemKeyConfig implements SslKeyConfig {
-    private final Path certificate;
-    private final Path key;
+
+    private static final String KEY_FILE_TYPE = "PEM private key";
+    private static final String CERT_FILE_TYPE = "PEM certificate";
+
+    private final String certificate;
+    private final String key;
     private final char[] keyPassword;
+    private final Path configBasePath;
 
-    public PemKeyConfig(Path certificate, Path key, char[] keyPassword) {
+    /**
+     * @param certificate    Path to the PEM formatted certificate
+     * @param key            Path to the PEM formatted private key for {@code certificate}
+     * @param keyPassword    Password for the private key (or empty is the key is not encrypted)
+     * @param configBasePath The base directory from which config files should be read (used for diagnostic exceptions)
+     */
+    public PemKeyConfig(String certificate, String key, char[] keyPassword, Path configBasePath) {
         this.certificate = Objects.requireNonNull(certificate, "Certificate cannot be null");
         this.key = Objects.requireNonNull(key, "Key cannot be null");
         this.keyPassword = Objects.requireNonNull(keyPassword, "Key password cannot be null (but may be empty)");
+        this.configBasePath = Objects.requireNonNull(configBasePath, "Config base path cannot be null");
+    }
+
+    @Override
+    public boolean hasKeyMaterial() {
+        return true;
     }
 
     @Override
     public Collection<Path> getDependentFiles() {
-        return Arrays.asList(certificate, key);
+        return Arrays.asList(resolve(certificate), resolve(key));
+    }
+
+    private Path resolve(String fileName) {
+        return configBasePath.resolve(fileName);
+    }
+
+    @Override
+    public Collection<? extends StoredCertificate> getConfiguredCertificates() {
+        final List<Certificate> certificates = getCertificates(resolve(this.certificate));
+        final List<StoredCertificate> info = new ArrayList<>(certificates.size());
+        boolean first = true;
+        for (Certificate cert : certificates) {
+            if (cert instanceof X509Certificate) {
+                info.add(new StoredCertificate((X509Certificate) cert, this.certificate, "PEM", null, first));
```

Comment:
Probably not - but that's what X-Pack currently does.

It _should_ be - if the certificate path points to a PEM file with multiple certificates, then they _should_ be a chain ordered from leaf to issuer to root - but there's no guarantee that it's true.

However, other things will break if it's not. When we build a keystore we will assume the certificates that we read form a valid chain, and `Keystore.setKeyEntry` will throw an exception if they're not.

Reply:
I added a test to show that `createKeyManager` fails if the chain is in the wrong order. I think that is sufficient for the assumption in this method.

We do have a "private key" for the leaf (first element) certificate. It might be the wrong key (none of the config classes check that in this method), but we have one.
And the core functionality of the config (building a key manager) will fail.